src/commonMain/kotlin/fr/acinq/lightning/serialization/v4/Serialization.kt

```diff
@@ -356,31 +357,68 @@ object Serialization {
         }
     }
 
+    private fun Output.writeUnsignedLocalCommitWithHtlcs(localCommit: InteractiveTxSigningSession.Companion.UnsignedLocalCommit) {
+        writeNumber(localCommit.index)
+        writeCommitmentSpecWithHtlcs(localCommit.spec)
+        writeTransactionWithInputInfo(localCommit.commitTx)
+        writeCollection(localCommit.htlcTxs) { writeTransactionWithInputInfo(it) }
+    }
+
+    private fun Output.writeLocalCommitWithHtlcs(localCommit: LocalCommit) {
+        writeNumber(localCommit.index)
+        writeCommitmentSpecWithHtlcs(localCommit.spec)
+        localCommit.publishableTxs.run {
+            writeTransactionWithInputInfo(commitTx)
+            writeCollection(htlcTxsAndSigs) { htlc ->
+                writeTransactionWithInputInfo(htlc.txinfo)
+                writeByteVector64(htlc.localSig)
+                writeByteVector64(htlc.remoteSig)
+            }
+        }
+    }
+
+    private fun Output.writeLiquidityPurchase(lease: LiquidityAds.Lease) {
+        writeNumber(lease.amount.toLong())
+        writeNumber(lease.fees.miningFee.toLong())
+        writeNumber(lease.fees.serviceFee.toLong())
+        writeByteVector64(lease.sellerSig)
+        writeNumber(lease.witness.fundingScript.size())
+        write(lease.witness.fundingScript.toByteArray())
+        writeNumber(lease.witness.leaseDuration)
+        writeNumber(lease.witness.leaseEnd)
+        writeNumber(lease.witness.maxRelayFeeProportional)
+        writeNumber(lease.witness.maxRelayFeeBase.toLong())
+    }
+
     private fun Output.writeInteractiveTxSigningSession(s: InteractiveTxSigningSession) = s.run {
         writeInteractiveTxParams(fundingParams)
         writeNumber(s.fundingTxIndex)
         writeSignedSharedTransaction(fundingTx)
         // We don't bother removing the duplication across HTLCs: this is a short-lived state during which the channel cannot be used for payments.
-        writeEither(localCommit,
```

Comment:
I think this conveys the intent better (NB: `writeEither` uses `write`, not `writeNumber`, for discriminators)

```kotlin
    private fun Output.writeInteractiveTxSigningSession(s: InteractiveTxSigningSession) = s.run {
        writeInteractiveTxParams(fundingParams)
        writeNumber(s.fundingTxIndex)
        writeSignedSharedTransaction(fundingTx)
        // We use 2/3 instead of the standard 0/1 as discriminator for the option to handle backward compatibility (this field was added later)
        when (liquidityPurchased) {
            null -> write(2)
            else -> {
                write(3)
                writeLiquidityPurchase(liquidityPurchased)
            }
        }
        // We don't bother removing the duplication across HTLCs: this is a short-lived state during which the channel cannot be used for payments.
        writeEither(localCommit,
            writeLeft = { writeUnsignedLocalCommitWithHtlcs(it) },
            writeRight = { writeLocalCommitWithHtlcs(it) }
        )
        remoteCommit.run {
            writeNumber(index)
            writeCommitmentSpecWithHtlcs(spec)
            writeTxId(txid)
            writePublicKey(remotePerCommitmentPoint)
        }
    }
```

Reply:
Right, nevermind!